pkg/route/ingress/ingress.go

```diff
@@ -635,7 +635,6 @@ func preserveRouteAttributesFromExisting(r, existing *routev1.Route) {
 	r.Spec.To.Weight = existing.Spec.To.Weight
 	if r.Spec.TLS != nil && existing.Spec.TLS != nil {
 		r.Spec.TLS.CACertificate = existing.Spec.TLS.CACertificate
-		r.Spec.TLS.DestinationCACertificate = existing.Spec.TLS.DestinationCACertificate
```

Comment:
We need to preserve this field for routes that don't have the annotation because if we don't, we will break any users who rely on the preservation of this field.  

Please add the following test case to make sure we don't break these users:
```go
		{
			name: "no-op - destination CA certificate has been changed by the user",
			fields: fields{
				i: &ingressLister{Items: []*networkingv1.Ingress{
					{
						ObjectMeta: metav1.ObjectMeta{
							Name:      "1",
							Namespace: "test",
							Annotations: map[string]string{
								"route.openshift.io/termination": "reencrypt",
							},
						},
						Spec: networkingv1.IngressSpec{
							TLS: []networkingv1.IngressTLS{
								{Hosts: []string{"test.com"}, SecretName: "secret-1"},
							},
							Rules: []networkingv1.IngressRule{
								{
									Host: "test.com",
									IngressRuleValue: networkingv1.IngressRuleValue{
										HTTP: &networkingv1.HTTPIngressRuleValue{
											Paths: []networkingv1.HTTPIngressPath{
												{
													Path:     "/",
													PathType: &pathTypePrefix,
													Backend: networkingv1.IngressBackend{
														Service: &networkingv1.IngressServiceBackend{
															Name: "service-1",
															Port: networkingv1.ServiceBackendPort{
																Name: "http",
															},
														},
													},
												},
											},
										},
									},
								},
							},
						},
						Status: networkingv1.IngressStatus{
							LoadBalancer: v1.LoadBalancerStatus{
								Ingress: []v1.LoadBalancerIngress{
									{Hostname: "apps.foo.com"},
								},
							},
						},
					},
				}},
				r: &routeLister{Items: []*routev1.Route{
					{
						ObjectMeta: metav1.ObjectMeta{
							Name:            "1-abcdef",
							Namespace:       "test",
							OwnerReferences: []metav1.OwnerReference{{APIVersion: "networking.k8s.io/v1", Kind: "Ingress", Name: "1", Controller: &boolTrue}},
							Annotations: map[string]string{
								"route.openshift.io/termination": "reencrypt",
							},
						},
						Spec: routev1.RouteSpec{
							Host: "test.com",
							Path: "/",
							To: routev1.RouteTargetReference{
								Name: "service-1",
							},
							Port: &routev1.RoutePort{
								TargetPort: intstr.FromString("http"),
							},
							WildcardPolicy: routev1.WildcardPolicyNone,
							TLS: &routev1.TLSConfig{
								Termination: routev1.TLSTerminationReencrypt,
								Key:                      "key",
								Certificate:              "cert",
								DestinationCACertificate: "cert",
							},
						},
						Status: routev1.RouteStatus{
							Ingress: []routev1.RouteIngress{
								{
									RouterCanonicalHostname: "apps.foo.com",
									Conditions: []routev1.RouteIngressCondition{{
										Type:   routev1.RouteAdmitted,
										Status: v1.ConditionTrue,
									}},
								},
							},
						},
					},
				}},
			},
			args: queueKey{namespace: "test", name: "1"},
		},
```

Then, make sure we still preserve the field in the annotation's absence:

```suggestion
		if _, ok := r.Annotations[destinationCACertificateAnnotationKey]; !ok {
			r.Spec.TLS.DestinationCACertificate = existing.Spec.TLS.DestinationCACertificate
		}
```

In addition, you will need to modify `routeMatchesIngress` likewise:

```diff
 	tlsConfig := tlsConfigForIngress(ingress, rule, tlsSecret, secretLister)
 	if route.Spec.TLS != nil && tlsConfig != nil {
 		tlsConfig.InsecureEdgeTerminationPolicy = route.Spec.TLS.InsecureEdgeTerminationPolicy
+		if _, ok := ingress.Annotations[destinationCACertificateAnnotationKey]; !ok {
+			tlsConfig.DestinationCACertificate = route.Spec.TLS.DestinationCACertificate
+		}
 	}
 	return reflect.DeepEqual(tlsConfig, route.Spec.TLS)
 }
```

Reply:
Added the test case and the suggestion. Github doesn't allow to apply this suggestion directly but is included in the PR.